server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateService.java

```diff
@@ -0,0 +1,371 @@
+                indexMetadata.getIndexUUID(),
+                indexMetadataKey
+            );
+            allUploadedIndexMetadata.add(uploadedIndexMetadata);
+        }
+        final ClusterMetadataMarker marker = uploadMarker(clusterState, allUploadedIndexMetadata, false);
+        final long durationMillis = relativeTimeMillisSupplier.getAsLong() - startTimeMillis;
+        if (durationMillis >= slowWriteLoggingThreshold.getMillis()) {
+            logger.warn(
+                "writing cluster state took [{}ms] which is above the warn threshold of [{}]; " + "wrote full state with [{}] indices",
+                durationMillis,
+                slowWriteLoggingThreshold,
+                allUploadedIndexMetadata.size()
+            );
+        } else {
+            // todo change to debug
+            logger.info(
+                "writing cluster state took [{}ms]; " + "wrote full state with [{}] indices",
+                durationMillis,
+                allUploadedIndexMetadata.size()
+            );
+        }
+        return marker;
+    }
+
+    /**
+     * This method uploads the diff between the previous cluster state and the current cluster state. The previous marker file is needed to create the new
+     * marker. The new marker file is created by using the unchanged metadata from the previous marker and the new metadata changes from the current cluster
+     * state.
+     *
+     * @return The uploaded ClusterMetadataMarker file
+     */
+    @Nullable
+    public ClusterMetadataMarker writeIncrementalMetadata(
+        ClusterState previousClusterState,
+        ClusterState clusterState,
+        ClusterMetadataMarker previousMarker
+    ) throws IOException {
+        final long startTimeMillis = relativeTimeMillisSupplier.getAsLong();
+        if (clusterState.nodes().isLocalNodeElectedClusterManager() == false) {
+            logger.error("Local node is not elected cluster manager. Exiting");
+            return null;
+        }
+        assert previousClusterState.metadata().coordinationMetadata().term() == clusterState.metadata().coordinationMetadata().term();
+        assert REMOTE_CLUSTER_STATE_ENABLED_SETTING.get(settings) == true : "Remote cluster state is not enabled";
+        final Map<String, Long> previousStateIndexMetadataVersionByName = new HashMap<>();
+        for (final IndexMetadata indexMetadata : previousClusterState.metadata().indices().values()) {
+            previousStateIndexMetadataVersionByName.put(indexMetadata.getIndex().getName(), indexMetadata.getVersion());
+        }
+
+        int numIndicesUpdated = 0;
+        int numIndicesUnchanged = 0;
+        final Map<String, ClusterMetadataMarker.UploadedIndexMetadata> allUploadedIndexMetadata = previousMarker.getIndices()
+            .stream()
+            .collect(Collectors.toMap(UploadedIndexMetadata::getIndexName, Function.identity()));
+        for (final IndexMetadata indexMetadata : clusterState.metadata().indices().values()) {
```

Comment:
Thinking about following scenario: Assume 
- `previousStateIndexMetadataVersionByName` contains indices `i1`, `i2`, `i3` and `i4`
- `allUploadedIndexMetadata` contains indices `i1` and `i2`
- `clusterState.metadata().indices()` contains `i2`, `i3`, `i4`, `i5` and `i6`

If `i3` is not changed between previous and current cluster state, we will skip uploading it.

Reply:
`previousStateIndexMetadataVersionByName` is created from `previousClusterState` and `allUploadedIndexMetadata` is created from `previousMarker`. `previousClusterState` and `previousMarker` should always be in sync since they set at the same time in the RemotePersistedState. So they will contain the same indices.

Reply:
In that case, do we need to get index metadata from `previousMarker` as well as `previousClusterState`? Why can't we rely on only one source?

Reply:
`previousMarker` is needed to know the details of `UploadedIndexMetadata` for all unchanged indices. This can be directly copied over to new marker. I have kept previousClusterState as in future, we can get the previous marker for remote store directly rather than keeping in memory.

Reply:
Previous marker contains the details of remote store blob keys. These are unique for every upload. So we need the previousMarker atleast. But I will add a TODO to revisit this logic.